Support border/padding/margin on RenderMathMLRow and subclasses

fred-wang opened the pull request:
#### 12803196d5e76b29a63777ce78612ba4e8ec15c1
<pre>
Support border/padding/margin on RenderMathMLRow and subclasses
<a href="https://bugs.webkit.org/show_bug.cgi?id=276227">https://bugs.webkit.org/show_bug.cgi?id=276227</a>

Reviewed by Rob Buis.

This implements support for border/margin/padding on the mrow, math,
mpadded, mroot, msqrt, menclose, merror, mphantom, mstyle elements
according to the rules described in MathML Core:

- When handling boxes of children during math layout, we consider *margin*
  boxes. Following other renderers, `recomputeLogicalWidth()` is called
  during the layout of children to set the inline margins (this does not
  seem enough for preferred width calculations though, but will be
  considered in follow-up patches) while during the parent we call
  `computeAndSetBlockDirectionMargins()` on each child in order to set
  set the block margins. To make the latter easier, the helper method
  RenderMathMLBlock::computeAndSetBlockDirectionMarginsOfChildren() is
  introduced.

- Current math layout is modified so that padding/border are added in
  order to obtain the border box. RenderMathMLRow already does that but
  then this can become wrong for derived classes (which add extra space
  or painting). Consequently, we modify RenderMathMLRow so the helpers
  getContentBoundingBox(), layoutRowItems() and
  stretchVerticalOperatorsAndLayoutChildren() do not actually consider
  the padding/border and it is up to the callers to add it around their
  own layout. Similarly, a new helper preferredLogicalWidthOfRowItems()
  is introduced to calculate the preferred logical width of a row of
  items without adding the border/padding. Finally we add a helper
  shiftRowItems() to shift children by an offset, something that
  was already used by some subclasses and can be used to shift by the
  top-left border/padding too.

Canonical link: <a href="https://commits.webkit.org/280733@main">https://commits.webkit.org/280733@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/007f170f8c1d686110abb8c50fcbf0f6206cda23

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57465 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36793 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9940 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61087 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7910 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8098 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46536 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5603 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34523 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49639 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27400 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31304 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6942 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6913 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53282 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62766 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1378 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53795 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1384 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49665 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53889 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12700 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1178 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32622 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33707 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34792 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33453 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis reviewed (approved):
LGTM. +s in commit title.

fred-wang committed:
Support border/padding/margin on RenderMathMLRow and subclasses
https://bugs.webkit.org/show_bug.cgi?id=276227

Reviewed by Rob Buis.

This implements support for border/margin/padding on the mrow, math,
mpadded, mroot, msqrt, menclose, merror, mphantom, mstyle elements
according to the rules described in MathML Core:

- When handling boxes of children during math layout, we consider *margin*
  boxes. Following other renderers, `recomputeLogicalWidth()` is called
  during the layout of children to set the inline margins (this does not
  seem enough for preferred width calculations though, but will be
  considered in follow-up patches) while during the parent we call
  `computeAndSetBlockDirectionMargins()` on each child in order to set
  set the block margins. To make the latter easier, the helper method
  RenderMathMLBlock::computeAndSetBlockDirectionMarginsOfChildren() is
  introduced.

- Current math layout is modified so that padding/border are added in
  order to obtain the border box. RenderMathMLRow already does that but
  then this can become wrong for derived classes (which add extra space
  or painting). Consequently, we modify RenderMathMLRow so the helpers
  getContentBoundingBox(), layoutRowItems() and
  stretchVerticalOperatorsAndLayoutChildren() do not actually consider
  the padding/border and it is up to the callers to add it around their
  own layout. Similarly, a new helper preferredLogicalWidthOfRowItems()
  is introduced to calculate the preferred logical width of a row of
  items without adding the border/padding. Finally we add a helper
  shiftRowItems() to shift children by an offset, something that
  was already used by some subclasses and can be used to shift by the
  top-left border/padding too.

Canonical link: https://commits.webkit.org/280733@main
